### ISSUE TITLE
Feature Request: Autocomplete Text

### ISSUE THREAD

**Issue by Born2Root** [pbek/QOwnNotes#2301]
**Feature Request on current behaviour**
It would be really good to have an text autocompletion feature included into QOwnNotes.
I know there is already one solution via scripts ( [Q] Request the auto-completion configure method #1817 , etc.)

But the currently way with hitting Crtl+Space all the time is not very useable to me. 
Especially because I don't know if the word I type is already in the autocomplete list or not, so at the moment its more an all the time pressing Ctrl+Space in contrast to a automatic working autocompletion.

**Expected Behaviour**
In my opinion a fullblown solution, that supports the complete dictionaries of several langauges is really not necessary.
Because that is the point, when performance will go down.

In Notepad++ they have a very nice autocomplete function, that I like very much.
It works just with the words in the current note file and the words from a user definded dictionary.
And well, thats already enough to safe you plenty of time while typing.

From the backend view, the function adds any word  with two or more characters long to a list that is used for the autocomplete.
And the user is able to configure the function, if it starts showing the autocompleting window starting from typed 2-5 letters.
So a hook could be, a word is typed in with the number of letters -> Reaction, show the autocompletion box, and the user is able to insert the word. If you don't insert and just type more the box vanishes or shows the more appropriate autocomplete words.
That process is automatic, so it's not necessary to hit Ctrl+Space to show the suggestion box.

I recommend to take a look at the solution how it is implemented in Notepad++ (preferably by trying it directly in Noetpad).
Otherwise theres an article at their documentation about the feature.

https://npp-user-manual.org/docs/auto-completion/
https://sodocumentation.net/notepadplusplus/topic/4843/auto-completion-and-suggestion
 
![notepad func](https://user-images.githubusercontent.com/8644900/137275630-e7a8b638-ee94-4563-9183-924a9e8673d6.JPG)

I would really appreciate it, if a similar funciton would be implemented in QOwnNotes.
Thats the only feature I'm currently missing, apart from that, the App is just brilliant.
Thanks at that point for your effort and the great work!

**Comment by pbek** [comment]
So basically we are talking about an "automatic ctrl+space", but only the "completing the current word by text found in the current note", right, @Born2Root?

Beside that I as user maybe wouldn't like to spend the cpu cycles to scan large files on every key-press it would also annoy me to have that box pop up automatically all the time...

What's your opinion, @Waqar144?

**Comment by pbek** [comment]
And it seems the QMenu, that is opened is stealing the focus, so you can't even type when it opens.

https://github.com/pbek/QOwnNotes/blob/96ddf262e3f7b7849c9daab3f86f0c618936c251/src/mainwindow.cpp#L10184

**Comment by Waqar144** [comment]
We can cache the word of a note so re-scanning may not be needed.

However, as a feature, I am not sure if its really needed because in general you don't have a lot of similar words in text and we aren't a code editor. It would also be very annoying when I am typing out my thoughts or taking notes quickly.

**Comment by pbek** [comment]
> We can cache the word of a note so re-scanning may not be needed.

But you'd need to re-scan as soon as something changes, wouldn't you?

> However, as a feature, I am not sure if its really needed because in general you don't have a lot of similar words in text and we aren't a code editor. It would also be very annoying when I am typing out my thoughts or taking notes quickly.

Yeah, in code editors it makes a lot of sense (when the code editor knows the syntax and command of the programming language to you are using)... But for notes...

What are you using QOwnNotes for, @Born2Root?

**Comment by Born2Root** [comment]
> _"So basically we are talking about an "automatic ctrl+space", but only the "completing the current word by text found in the current note", right, @Born2Root?"_

Exactly especially when typing technical terms that safes me a lot of time.
First of all in other languages than english, german for example, because the words are sometimes pretty long there.

> _"What are you using QOwnNotes for, @Born2Root?"_

I use it for note taking and writing technical essays and BlogPosts (mega handy because I can use markdown).
Primarily there I have many technical termns that are longer or have a very own capitalization.
Like in the term "QOwnNotes" with that camelcase like writing or terms like SecuFEx or Drive-CLiQ, etc.
And because I'm storing the BlogPosts in the same notefile, the terms and words that I use are repeated quite often.

Especially in writing the longer texts I find QOwnNotes really handy.
Because with the window management and the distraction free mode I can switch easily between note taking and writing.
And apart fromt that I have all my stuff nicely sorted in the same tool, under the same interface and with cross referenced notes.

At the moment my workaround is, that I write the long texts in Notepad++ and copy them over to QOwnNotes afterwards...

**Comment by Waqar144** [comment]
> But you'd need to re-scan as soon as something changes, wouldn't you?

Not after every change, that would be overkill, but after intervals or every new line entered. However this is the easy part.

Automatic invocation is the harder part in this. We need to figure out when to invoke, e.g after n number of characters typed, and when to close the completion box. And other things like filtering the items based on the chars typed and positioning/sizing of the box itself. And we can have multiple sources of data i.e., dictionaries, current document so handling and merging their input. 

In general, I have never seen any word processors/text fields do this.

**Comment by Born2Root** [comment]
Like I mentioned above, just check it out on Notepad++ yourself to get an impression how handy it could be.
Not just in coding, but also in fast writing texts.

[https://notepad-plus-plus.org/](url)

Documentation about the feature:
https://npp-user-manual.org/docs/auto-completion/
https://sodocumentation.net/notepadplusplus/topic/4843/auto-completion-and-suggestion

**Comment by pbek** [comment]
Even if I'd like the idea, there still currently is no way (that I know of) to have an auto-completion that doesn't take the focus while typing in a QPlainTextEdit...

**Comment by Waqar144** [comment]
It is possible of course, we just need to play around a bit with the focus policy. 

@Born2Root thanks for the input. My point was that this feature then is no longer just a simple method that does a small-ish task (which is what we currently have). A proper framework to handle completion will be needed which wont be just a function but a few classes that implement all the needed functionality. Thus, imo, this falls beyond the scope of what QON is, i.e., a note taker.

**Comment by Born2Root** [comment]
Ok, understood and I think that's a reasonable decision.
Still, thank you very much for the fast responses and the very solution orientated discussion.

**Comment by ferrlen** [comment]
I'm sorry, I hope it's alright if I add a comment to a closed threat. But although QQOwnNotes blows my mind in pretty much every way, I also end up using Notepad++ to write long texts because I find it faster (it also helps me to locate myself in bigger files having numbered lines). I then copy and paste it to QQOwnNotes for advanced features, or use QQOwnNotes for short notes.

Unfortunately, I can't code in C++, but from what I understood, to add the possibility of an [automatic Ctrl+space that only scans the current document for words](https://github.com/pbek/QOwnNotes/issues/2301#issuecomment-971634685) one would need to stop the QMenu getting focus when opened and open the QMenu at every letter stroke?

If one changes QMenu behaviour to only get focus when `tab` or `Enter` is pressed and to close otherwise, would it be easy to open QMenu at every letter input? Changing it to only steal focus like this might be a good idea anyway, if the user presses `Ctrl+space` but gives up on using QMenu, one needs to press `Esc` to return to normal. 

I totally agree with you that it'd be annoying for users to always have QMenu opening, but for folks like us being able to turn on an option of triggering QManu at every QStringLiteral would be super useful. Would it be easy for a newbie like me to write a QQOwnNotes script implementing this idea? I'm comfortable with coding in Javascript

Thanks a lot!

* * *
TL;DR I'd like to create a *toggle* option to invoke QMenu at every letter input. Can I write a [script](https://github.com/pbek/QOwnNotes/blob/develop/docs/scripting/examples/autocompletion.qml) to:
- make QMenu only steal focus if `tab` or `Enter` are pressed
- open QMenu if letters are typed

**Comment by pbek** [comment]
Yes, but beside having to implement a new menu widget and creating a real framework for auto-completion I still think that's a little out of scope. 😅  see https://github.com/pbek/QOwnNotes/issues/2301#issuecomment-972665955
